Add Route movement and related methods

meshy opened the pull request:
Deferred from #75:
- [x] Rebase once #75 is merged.
- [x] Add `Route.objects.move_branch(old_url, new_url)`.
- [x] Add `Route().get_ancestors()`
- [x] Add `Route().swap_with(other_route, move_children)`
- [x] Add `Route().move_to(new_url, move_children)`
- [x] Decide if these method ought to be in atomic transactions or not. *Conclusion: not at the moment.*
- [x] Determine if validation is worthwhile on the above methods. *Conclusion: yes, but optionally disabled with flag. Defer to another PR.*
- [x] Raise error from `get_ancestors` when called on object that hasn't yet been saved to the database (has no `pk`).
- [x] Mark user-facing strings as translatable.
- [x] Consider re-writing long queries to clarify their single-query nature (remove line-smearing variables).

Trashed ideas:

~~- [ ] Add `Route.objects.swap_urls(first, second)` method.~~
~~- [ ] Add `Route.objects.swap_branches(first, second)` method.~~
~~- [ ] Add `Route().get_parent()`~~

meshy commented on conman/routes/managers.py:
```diff
@@ -30,3 +32,31 @@ def best_match_for_path(self, path):
+        tmp_path = str(uuid.uuid4())
+        self.filter(url=path_a).update(url=tmp_path)
+        self.filter(url=path_b).update(url=path_a)
+        self.filter(url=tmp_path).update(url=path_b)
```
Would it be better to do a `get`, and raise errors when `Routes` don't exist? If so, should we be doing this in an atomic transaction?

Would [`select_for_update`](https://docs.djangoproject.com/en/1.9/ref/models/querysets/#select-for-update) be worth considering to avoid race conditions? Is that a threat?

KevinEtienne replied:
Not sure about the `select_for_update` but the atomic transaction seems a good idea.

meshy replied:
Thanks `:)` as detailed in [another comment](https://github.com/meshy/django-conman/pull/80#discussion_r69457152), I think it might end up going that way.

Ian-Foote commented on conman/routes/managers.py:
```diff
@@ -30,3 +32,31 @@ def best_match_for_path(self, path):
+
+    def swap_urls(self, path_a, path_b):
+        """
+        Swap the urls of two Routes without moving their descendants.
```
Not clear to me. This wouldn't significantly change the code here (if I understand correctly) so it's much more important how these are going to be used. Should the paths be considered implementation details of Routes in this context or are they really the core concept here?

meshy replied:
> Should the paths be considered implementation details of Routes in this context or are they really the core concept here?

I'd say they're key to the concept, really.

> it's much more important how these are going to be used.

Yeah, that's the angle I was coming from. I'm primarily concerned with making the API logical.

I'm expecting to make use of these methods when I come to implementing custom admin actions/views, but would also like to make them accessible to users of the library.

meshy replied:
Further thoughts:

> This wouldn't significantly change the code here (if I understand correctly)

I'm beginning to think that the `update` method employed here is not a good idea, and that I should instead be doing a `SELECT` to get the `Routes` followed by a `save()`. ([I've alluded to this in another comment.](https://github.com/meshy/django-conman/pull/80#discussion_r69424057)) This would allow two things:

- We could raise errors when asked to swap things that do not exist.
- We might add a hook for custom logic/validation.

This second point springs from an idea I have for allowing `Routes` to have  dominion over sub-URLs. ie: the ArticleRoute at `/articles/` would be invoked when the `/articles/2016-07-04-flag-waving` url is hit, and, as such, might dictate that having child urls is disallowed.

Disallowing children might well make this significantly more complex, but is a concept I hope to introduce (unsure if it should be core...probably though).

meshy replied:
Further further thoughts:

- In hindsight, disallowing children may b e unwise. I can imagine a whole ton of edge cases being hard to manage, and it also means that views with dynamic subroutes could not be overridden if need be.

meshy replied:
My current thinking:

Add `Route().swap_with(other_route, *, move_children)`.

When `move_children` is `False`, this method is called.

When `move_children` is `True`, we'll call `move_branch` three times, similar to how we call `swap_urls`.

Ian-Foote commented on conman/routes/managers.py:
```diff
@@ -30,3 +32,31 @@ def best_match_for_path(self, path):
+
+    def move_branch(self, old_url, new_url):
+        """
+        Move this Route and all descendants to a new URL.
```
If this matches your intention here, I'm surprised this is a `Manager` method and not a `Model` method.

meshy replied:
Good catch, ta. This was previously a model method, and I moved it. Must have forgotten to update the docstring!

meshy replied:
I've been thinking of adding a `Route().move_to(new_url, *, descendants=False)` method.

When `descendants` is `True`, it'll call this method, and when `False`, it'll just set `url`, and attempt to save.

Sound reasonable?

meshy replied:
Or do you think I should just fold this method into the `move_to` method?

meshy commented:
@KevinEtienne @Ian-Foote thanks for the feedback guys, but it's probably not worth reviewing this yet, as I'm still working on it, and #75 needs to be merged in first.

meshy commented:
(I very much appreciate the feedback, BTW `:)`)

meshy commented:
Oh, just realised what was going on there... I was asking questions of myself to answer later! Should have known better! Glad to have your answers, and now feel bad for stopping you! `:P`

KevinEtienne commented on CHANGELOG.md:
```diff
@@ -8,6 +8,14 @@
+
+* `Route.objects.move_branch()` can be used to move a `Route` and its
+  descendants to another location.
+* `Route.objects.swap_urls()` can be used to swap the locations of two
```
should you specify the arguments for `move_branch` and `swap_urls`?

meshy commented on conman/routes/models.py:
```diff
@@ -82,6 +82,27 @@ def handle(self, request, path):
         # Deal with the request
         return handler.handle(request, path)
 
+    def move_to(self, new_url, *, move_children):
```
Not sure about the phrase `move_children`... What do you think, @Ian-Foote and @KevinEtienne? Perhaps this ought to be `descendants`? Is there a better keyword you can think of?

KevinEtienne replied:
I see what you mean but I think it's clear enough when you read the code. Otherwise here are a few suggestions I could think of are but they are all composed with two words:
 - `move_to(self, new_url, *, include_children)`;
 - `move_to(self, new_url, *, affect_children)`;
 - `move_to(self, new_url, *, update_subpath)`.

Ian-Foote commented on conman/routes/utils.py:
```diff
@@ -13,7 +13,7 @@ def split_path(path):
+        paths.insert(1, path + '/')
         path = os.path.split(path)[0]
         if path == '/':
             break
```
It probably doesn't matter much, but if we wanted to optimise this, we could try using a `collections.deque` and the `appendleft` method. The advantages might not appear with typical length urls though, so profiling would be a good idea.

meshy replied:
I doubt profiling would reveal any noticeable difference between this and a `deque`, but I suspect it would be a whole lot easier to read. I'll pop a `deque` in there, thanks for the suggestion `:)`

meshy replied:
Got a pretty nice refactor of this now -- incoming.

meshy replied:
https://github.com/meshy/django-conman/pull/80/commits/4fd1713043d736b4bcc5defc9224a2291c480c74

Ian-Foote commented on conman/routes/models.py:
```diff
@@ -74,6 +84,64 @@ def handle(self, request, path):
+                self.save()
+            except IntegrityError:
+                self.url = old_url
+                raise
```
Aha! I just noticed your open question in the top comment. My answer is yes, these methods should be atomic.

meshy replied:
Yeah, I was leaning that way too. I'll `transaction.atomic` them.

meshy replied:
Come to think of it, given this method only ever executes one query, you could argue that it's already atomic. The only difference is that of not automatically allowing further queries after the  `IntegrityError`, which may not always be required.

I'll add `atomic` to `swap_with`, as that does multiple queries, but unless you think it's important that this changes, I'll leave it as is.

meshy replied:
Come to think of it, `swap_with` is intended to be pretty safe. I can (just about) imagine how one might make it fail, but it's such an odd set of circumstances that I don't expect it to ever happen deliberately. Making it atomic seems wasteful, given any problem ought to be caught by `ATOMIC_REQUESTS`, and a 500 raised.

Ian-Foote replied:
I don't think we can rely on `ATOMIC_REQUESTS` being set.

meshy replied:
Realistically we can't, but if it's off, then people should be used to managing the fallout from errors being raised. Seems like something django ought to have a `check` for IMO, but failing that, perhaps we should add one to this project.

meshy replied:
I think I might be too tired to make a sensible decision on this at the moment, despite having had a change of mind about this. Perhaps it'd be a good one to discuss in person.

meshy commented on conman/routes/managers.py:
```diff
@@ -30,3 +30,17 @@ def best_match_for_path(self, path):
+        to_move = self.filter(url__startswith=old_url)
+        to_move.update(url=Concat(
+            Value(new_url),
+            Substr('url', len(old_url) + 1),  # 1 indexed
```
Thanks for noticing! This is one of my favourite parts of the whole codebase `:D`

meshy commented on conman/routes/models.py:
```diff
@@ -74,6 +84,64 @@ def handle(self, request, path):
         # Deal with the request
         return handler.handle(request, path)
 
+    def move_to(self, new_url, *, move_children):
```
I suspect either default might lead to cases of unexpected pain, so I thought it might be best to force the programmer to make a decision one way or the other, given the naming doesn't clearly suggest a default. Do you think this might be clunky as is?

meshy commented on conman/routes/models.py:
```diff
@@ -74,6 +84,64 @@ def handle(self, request, path):
+        See https://code.djangoproject.com/ticket/20581.
+        """
+        if not (self.pk and other_route.pk):
+            raise ValueError('Cannot move unsaved Routes.')
```
I'm not sure of a better alternative, but I'm open to suggestion?

meshy replied:
I suspect we're talking about two situations here.

1. Model has been instantiated, but never saved to the database. This is what I intended to protect against. In this case, if someone has explicitly set `.pk` on the model... well... they're daft, otherwise, I think checking `pk` should be fine.

2. Model has been saved at some point, but `.url` doesn't match what's in the database. I think this is where `django-dirtyfields` might come in handy, but that seems like a fair bit of overhead to add to protect against this behaviour. I think (...hope) that it's unlikely that someone would edit the `url` manually, fail to call `save`, and the pass it into `swap_with`.

meshy replied:
That said, I now think the error message should be more specific `:)`.

meshy commented on conman/routes/models.py:
```diff
@@ -74,6 +84,64 @@ def handle(self, request, path):
+            Route.objects.move_branch(self.url, tmp_path)
+            Route.objects.move_branch(other_route.url, self.url)
+            Route.objects.move_branch(tmp_path, other_route.url)
+            other_route.url, self.url = self.url, other_route.url
```
`move_branch` will update these rows in the database, but this gives us a chance to update the values of these objects without a call to `refresh_from_db` (which would be required, as `move_branch` uses `Route.objects.update()`).

meshy commented on conman/routes/tests/test_models.py:
```diff
@@ -44,6 +44,71 @@ def test_unique_url(self):
+
+        with self.assertNumQueries(0):
+            # Presumed nonsense as unsaved, so no query.
+            ancestors = list(route.get_ancestors())
```
In hindsight, yes, I think that would be more consistent.

meshy replied:
Any ideas what the correct exception might be?

Ian-Foote replied:
One possibility is [`FieldError`](https://docs.djangoproject.com/en/1.10/ref/exceptions/#django.core.exceptions.FieldError), since the problem is the primary key is not set.

Another possibility is that `assert self.pk is not None` would be sufficient - is it reasonable to expect this will only occur if the programmer using conman has made a mistake?

Otherwise, perhaps one of `TypeError`, `ValueError` or `RuntimeError`, but I'm not thrilled by these.

I'm currently leaning towards the `assert`, but I can be convinced otherwise.

Ian-Foote replied:
You do use `ValueError` elsewhere for a similar check, so that's a point in its favour.

meshy replied:
I've changed the "similar check" to use `assert self.pk`, and have used the `assert` here too. Thanks `:)`

meshy commented on conman/routes/managers.py:
```diff
@@ -30,3 +30,17 @@ def best_match_for_path(self, path):
+
+    def move_branch(self, old_url, new_url):
+        """
+        Move a Route and all descendants to a new URL.
```
Come to think of it... it might be worth considering some kind of validation on these, otherwise I might be making a bit of a foot-gun here.

meshy replied:
d03a68402e15affc3037e83e5e9b1d58460b518c

Ian-Foote commented on conman/routes/utils.py:
```diff
@@ -8,16 +8,20 @@ def split_path(path):
-        if path == '/':
-            break
-    return paths
+        path = path.rpartition('/')[0]
```
Ooh, I didn't think of using `rpartition`.

meshy replied:
haha don't feel bad -- this is the third time I've written this function, and it's only just occurred to me `;)`

meshy commented on conman/routes/models.py:
```diff
@@ -74,6 +84,71 @@ def handle(self, request, path):
+        See https://code.djangoproject.com/ticket/20581.
+        """
+        if not (self.pk and other_route.pk):
+            raise ValueError('Cannot move unsaved Routes.')
```
This should be `_('translatable')`

meshy commented on conman/routes/models.py:
```diff
@@ -74,6 +84,71 @@ def handle(self, request, path):
+
+        urls = sorted((self.url, other_route.url))
+        if move_children and urls[1].startswith(urls[0]):
+            msg = 'Cannot move children when swapping ancestors with descendants.'
```
`_('translatable')`

meshy commented:
Ok, I think I've addressed everything. I've refactored pretty heavily to reduce churn in the history.

@Ian-Foote might I ask for another look and (if you're happy) a 👍

Ian-Foote commented on conman/routes/models.py:
```diff
@@ -45,8 +46,9 @@ def check(cls):
     def get_ancestors(self):
         """Get all the ancestors of this Route."""
-        paths = split_path(self.url)[:-1]
+        assert self.pk  # Ensure object has been saved.
```
Extreme edge-case - can `pk` ever be `0` (or other falsey value)?

meshy replied:
By the looks of it, [it can be explicitly set to `0`](https://github.com/django/django/commit/d3cf6cfacfb828faad4f4f97c904e259304649b3)!

I have no idea why someone would want to do that, but I'll update this to deal with it. Good catch!

meshy commented:
`\o/` Thanks! `:)`
